go/core/action.go

```diff
@@ -28,6 +28,38 @@ import (
 	"github.com/invopop/jsonschema"
 )
 
+// Middleware is a function that takes in an action handler function and
+// returns a new handler function that might be changing input/output in
+// some way.
+//
+// Middleware functions can:
+//   - execute arbitrary code;
+//   - change the request and response;
+//   - terminate response by returning a response (or error);
+//   - call the next middleware function.
+type Middleware[I, O any] func(MiddlewareHandler[I, O]) MiddlewareHandler[I, O]
+
+type MiddlewareHandler[I, O any] func(ctx context.Context, input I) (O, error)
+
+// Middlewares returns an array of middlewares that are passes in as an argument.
+// core.Middlewares(apple, banana) is identical to []core.Middleware[InputType, OutputType]{apple, banana}
```

Comment:
I'm not sure why people wouldn't just use a slice constructor here.  Does this function pull its weight?

Reply:
So, for example here's what it will look like when trying to pass in model middleware:
```go
[]core.Middleware[*ai.GenerateRequest, *ai.GenerateResponse]{m1, m2}
```

vs
```go
core.Middlewares(m1, m2)
```

I prefer the second one...

Reply:
Ah, no type inference on compound literals.  Fair point.